sys/senml/senml.c

```diff
@@ -0,0 +1,325 @@
+            }
+            else if (!strcmp(tmp_key, SJ_VALUE_SUM)) {
+                (pack->records)[array_num].value_sum = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_TIME)) {
+                (pack->records)[array_num].time = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_UPDATE_TIME)) {
+                (pack->records)[array_num].update_time = atof(tmp_val);
+            }
+        }
+    }
+
+    pack->num = (array_num + 1);
+
+    return 0;
+}
+
+
+int senml_encode_json_s(const senml_pack_t *pack, char *output, size_t len)
+{
+    size_t insert_pos = 0;
+
+    output[0] = '[';
+    insert_pos++;
+    len--;
+
+    int chars_written;
+
+    if (!pack->base_info) {
+        goto encode_records;
+    }
+
+    output[insert_pos] = '{';
+    insert_pos++;
+    len--;
+
+    // using len - 2 to leave space for closing } and ]
+
+    if (pack->base_info->version) {
+        chars_written = snprintf(&output[insert_pos], len - 2, "\"bver\":%u,",
+                                 pack->base_info->version);
+
+        insert_pos += chars_written;
+        len -= chars_written;
+
+        if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+            return -1;
+        }
+    }
+
+    if (pack->base_info->base_name) {
+        chars_written = snprintf(&output[insert_pos], len - 2, "\"bn\":\"%s\",",
+                                 pack->base_info->base_name);
+
+        insert_pos += chars_written;
+        len -= chars_written;
+
+        if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+            return -1;
+        }
+    }
+
+    if (pack->base_info->base_time != 0) {
+        chars_written = snprintf(&output[insert_pos], len - 2, "\"bt\":%f,",
+                                 pack->base_info->base_time);
+
+        insert_pos += chars_written;
+        len -= chars_written;
+
+        if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+            return -1;
+        }
+    }
+
+    if (pack->base_info->base_unit) {
+        chars_written = snprintf(&output[insert_pos], len - 2, "\"bu\":\"%s\",",
+                                 pack->base_info->base_unit);
+
+        insert_pos += chars_written;
+        len -= chars_written;
+
+        if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+            return -1;
+        }
+    }
+
+    if (pack->base_info->base_value != 0) {
+        chars_written = snprintf(&output[insert_pos], len - 2, "\"bv\":%f,",
+                                 pack->base_info->base_value);
+
+        insert_pos += chars_written;
+        len -= chars_written;
+
+        if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+            return -1;
+        }
+    }
+
+    encode_records:
+
+    for (size_t i = 0; i < pack->num; i++) {
+        senml_record_t *curr_record = &(pack->records[i]);
+
+        if (i != 0) {
+            output[insert_pos] = '{';
```

Comment:
Should not be an issue anymore, since the function will return -1 if cnt_len would be 0 after a call to snprintf